README.md
pygtail
=======

A python "port" of [logcheck's logtail2](http://logcheck.org).

Pygtail reads log file lines that have not been read. It will even handle log
files that have been rotated.

Usage
-----

From the command line:

    Usage: pygtail.py [options] logfile

    Print log file lines that have not been read.

    Options:
      -h, --help            show this help message and exit
      -o OFFSET_FILE, --offset-file=OFFSET_FILE
                            File to which offset data is written (default:
                            <logfile>.offset).
      -p, --paranoid        Update the offset file every time we read a line
                            (as opposed to only when we reach the end of the
                            file).
      -n N, --every-n=N     Update the offset file every N'th time we read a
                            line (as opposed to only when we reach the end of
                            the file).
      --no-copytruncate     Don't support copytruncate-style log rotation.
                            Instead, if the log file shrinks, print a warning.
      --read-from-end       Read log file from the end if offset file is
                            missing. Useful for large files.
      --log-pattern         Custom log rotation glob pattern. Use %s to
                            represent the original filename. You may use this
                            multiple times to provide multiple patterns.
      --full_lines          Only log when line ends in a newline `\n`
                            (default: False)
      --encoding ENCODING   Encoding to use for reading files (default: system
                            encoding)
      --version             Print version and exit.

In your code:

```python
from pygtail import Pygtail

for line in Pygtail("some.log"):
    sys.stdout.write(line)
```

An example showing iterating over lines with offsets and manual control over offset updates:

```python
import pygtail

tail = pygtail.Pygtail(logfile, save_on_end=False, copytruncate=False)

for line, offset in tail.with_offsets():
    # Do someting

 # figure out right offset to save
 tail.write_offset_to_file(right_offset)
```

Contributing
------------

Pull requests are very much welcome, but I will not merge your changes if you don't include a test. Run tests with `python setup.py test`.

Build status
------------

[![Build Status](https://secure.travis-ci.org/bgreenlee/pygtail.png)](http://travis-ci.org/bgreenlee/pygtail)




debian/changelog
pygtail (0.14.0) lucid; urgency=low

  * Added with_offsets iterator and write_offset_to_file method. (thanks @pajowu!)
  * Included LICENSE file in distribution.

 -- Brad Greenlee <brad@footle.org>  Sun, 6 Nov 2022 06:48:00 -0800

pygtail (0.13.0) lucid; urgency=low

  * Add option to set the file encoding. (thanks @Robert-Lebedeu and @ondrej-bouda!)
  * Add test for renamecreate when file is rotated to unknown name and pygtail is freshly reading file (thanks @arekm!)
  * Add support for failing test proposed in #42

 -- Brad Greenlee <brad@footle.org>  Sun, 30 Oct 2022 11:48:00 -0700

 pygtail (0.12.0) lucid; urgency=low

  * Add option to disable save on end. (thanks @pajowu!)

 -- Brad Greenlee <brad@footle.org>  Thu, 14 Apr 2022 08:31:00 -0700

pygtail (0.11.1) lucid; urgency=low

  * Update debian packaging

 -- Sebastien GALLET <bibi21000@gmail.com>  Mon, 11 Oct 2019 11:30:00 +0100

pygtail (0.11.0) lucid; urgency=low

  * Add option to only log lines ending in a newline (thanks @MDunitz!)

 -- Brad Greenlee <brad@footle.org>  Tue, 16 Jul 2019 05:12:00 -0700

pygtail (0.10.1) lucid; urgency=low

  * Fix keyword argument mismatch (thanks @ushuz!)

 -- Brad Greenlee <brad@footle.org>  Mon, 04 Mar 2019 22:18:00 -0800

pygtail (0.10.0) lucid; urgency=low

  * Add --log-pattern flag to allow for custom rotated log filename patterns. (thanks @silenben)

 -- Brad Greenlee <brad@footle.org> Tue, 20 Nov 2018 04:48:00 -0800

pygtail (0.9.0) lucid; urgency=low

  * Add --read-from-end flag to read the log file from the end if the offset file is missing. (thanks Dmitry Sytsko)

 -- Brad Greenlee <brad@footle.org> Thu, 02 Aug 2018 16:52:00 -0700

pygtail (0.8.0) lucid; urgency=low

  * Add handling for "renamecreate" semantics where the currently processed file gets renamed and the
    original file gets recreated. This is the behavior of certain logfile rollers such as
    TimeBasedRollingPolicy in Java's Logback library. (thanks @bobtiernay-okta)

 -- Brad Greenlee <brad@footle.org>  Sat, 10 Mar 2018 08:39:18 -0800

pygtail (0.7.0) lucid; urgency=low

  * Add handling for logrotate dateext dateformat -%Y%m%d-%s log files, with and without
    delaycompress (thanks @akashawasthi)

 -- Brad Greenlee <brad@footle.org>  Thu, 30 Jun 2016 20:25:05 -0700

pygtail (0.6.1) lucid; urgency=low

  * Enable line buffering to avoid reading incomplete lines (thanks @olystretch)

 -- Brad Greenlee <brad@footle.org>  Thu, 05 Nov 2015 20:13:52 -0800

pygtail (0.6.0) lucid; urgency=low

  * Added flag to write offset file only after every n lines, and ability to optionally call a
    function whenever the offset file is written (thanks @btilly)

 -- Brad Greenlee <brad@footle.org>  Thu, 08 Oct 2015 05:19:19 -0700

pygtail (0.5.3) lucid; urgency=low

  * Fixed Python 3 compatibility (thanks @sbraz)

 -- Brad Greenlee <brad@footle.org>  Thu, 16 Apr 2015 17:30:43 -0700

pygtail (0.5.2) lucid; urgency=medium

  * Fixed issue with potentially incorrect offset due to python's read-ahead buffer (thanks @czchen)

 -- Brad Greenlee <brad@footle.org>  Thu, 26 Mar 2015 09:10:36 -0400

pygtail (0.5.1) lucid; urgency=low

  * Fixed print statement to be Python 3 compatible (thanks @benburry)

 -- Brad Greenlee <brad@footle.org>  Fri, 4 Jan 2015 20:36:00 -0800

pygtail (0.5.0) lucid; urgency=low

  * Add handling of rotated and immediately compressed files (thanks @NotSqrt)

 -- Brad Greenlee <brad@footle.org>  Fri, 17 Oct 2014 09:39:00 -0700

pygtail (0.4.0) lucid; urgency=low

  * Make support for copytruncate log rotation the default.
  * Add --no-copytruncate commandline switch.
  * Print warning if copytruncate is disabled and the log file shrinks

 -- Brad Greenlee <brad@footle.org>  Fri, 9 May 2014 17:31:00 -0700

pygtail (0.3.0) lucid; urgency=low

  * Add support for copytruncate style logrotation (contributed by https://github.com/atward)

 -- Brad Greenlee <brad@footle.org>  Fri, 9 May 2014 16:00:00 -0700

pygtail (0.2.2) lucid; urgency=low

  * Add parsing support for python's TimedRotatingFileHandler (contributed by https://github.com/darknight)

 -- Brad Greenlee <brad@footle.org>  Wed, 21 Aug 2013 15:50:00 -0700

pygtail (0.2.1) lucid; urgency=low

  * Add ReStructuredText README for PyPi.

 -- Brad Greenlee <brad@footle.org>  Wed, 06 Jul 2011 12:37:08 -0700

pygtail (0.2) lucid; urgency=low

  * Add setup.py and debianize.

 -- Brad Greenlee <brad@footle.org>  Wed, 05 Jul 2011 15:04:42 -0700


pygtail_test.py
import os
import unittest
import tempfile
from pygtail import Pygtail

class PygtailTest(unittest.TestCase):
    # TODO:
    # - test for non-default offset file
    # - test for paranoid flag
    # - test for savelog and datext rotation schemes

    def setUp(self):
        self.test_lines = ["1\n", "2\n", "3\n"]
        self.test_str = ''.join(self.test_lines)
        self.logfile = tempfile.NamedTemporaryFile(delete=False)
        self.logfile.write(self.test_str)
        self.logfile.close()

    def append(self, str):
        # append the give string to the temp logfile
        fh = open(self.logfile.name, "a")
        fh.write(str)
        fh.close()

    def tearDown(self):
        filename = self.logfile.name
        for tmpfile in [filename, filename + ".offset", filename + ".1"]:
            if os.path.exists(tmpfile):
                os.remove(tmpfile)

    def test_read(self):
        pygtail = Pygtail(self.logfile.name)
        self.assertEquals(pygtail.read(), self.test_str)

    def test_readlines(self):
        pygtail = Pygtail(self.logfile.name)
        self.assertEquals(pygtail.readlines(), self.test_lines)

    def test_subsequent_read_with_no_new_data(self):
        pygtail = Pygtail(self.logfile.name)
        self.assertEquals(pygtail.read(), self.test_str)
        self.assertEquals(pygtail.read(), None)

    def test_subsequent_read_with_new_data(self):
        pygtail = Pygtail(self.logfile.name)
        self.assertEquals(pygtail.read(), self.test_str)
        new_lines = "4\n5\n"
        self.append(new_lines)
        new_pygtail = Pygtail(self.logfile.name)
        self.assertEquals(new_pygtail.read(), new_lines)

    def test_logrotate(self):
        new_lines = ["4\n5\n", "6\n7\n"]
        pygtail = Pygtail(self.logfile.name)
        pygtail.read()
        self.append(new_lines[0])
        os.rename(self.logfile.name, "%s.1" % self.logfile.name)
        self.append(new_lines[1])
        pygtail = Pygtail(self.logfile.name)
        self.assertEquals(pygtail.read(), ''.join(new_lines))

def main():
    unittest.main()


if __name__ == "__main__":
    main()

pygtail.py
#!/usr/bin/python -tt
# -*- coding: utf-8 -*-

# pygtail - a python "port" of logtail2
# Copyright (C) 2011 Brad Greenlee <brad@footle.org>
#
# Derived from logcheck <http://logcheck.org>
# Copyright (C) 2003 Jonathan Middleton <jjm@ixtab.org.uk>
# Copyright (C) 2001 Paul Slootman <paul@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from os import stat
from os.path import exists
import sys
import glob
import string
from optparse import OptionParser

class Pygtail(object):
    """
    Creates an iterable object that returns only unread lines.
    """
    def __init__(self, filename, offset_file=None, paranoid=False):
        self.filename = filename
        self.paranoid = paranoid
        self._offset_file = offset_file or "%s.offset" % self.filename
        self._offset_file_inode = 0
        self._offset = 0
        self._fh = None
        self._rotated_logfile = None

        # if offset file exists, open and parse it
        if exists(self._offset_file):
            offset_fh = open(self._offset_file, "r")
            (self._offset_file_inode, self._offset) = \
                [string.atoi(line.strip()) for line in offset_fh]
            offset_fh.close()
            if self._offset_file_inode != stat(self.filename).st_ino:
                # The inode has changed, so the file might have been rotated.
                # Look for the rotated file and process that if we find it.
                self._rotated_logfile = self._determine_rotated_logfile()

    def __iter__(self):
        return self

    def next(self):
        """
        Return the next line in the file, updating the offset.
        """
        try:
            line = self._filehandle().next()
        except StopIteration:
            # we've reached the end of the file; if we're processing the
            # rotated log file, we can continue with the actual file; otherwise
            # update the offset file
            if self._rotated_logfile:
                self._rotated_logfile = None
                self._fh.close()
                self._offset = 0
                # open up current logfile and continue
                try:
                    line = self._filehandle().next()
                except StopIteration: # oops, empty file
                    self._update_offset_file()
                    raise
            else:
                self._update_offset_file()
                raise

        if self.paranoid:
            self._update_offset_file()

        return line

    def readlines(self):
        """
        Read in all unread lines and return them as a list.
        """
        return [line for line in self]

    def read(self):
        """
        Read in all unread lines and return them as a single string.
        """
        lines = self.readlines()
        if lines:
            return ''.join(lines)
        else:
            return None

    def _filehandle(self):
        """
        Return a filehandle to the file being tailed, with the position set
        to the current offset.
        """
        if not self._fh or self._fh.closed:
            filename = self._rotated_logfile or self.filename
            self._fh = open(filename, "r")
            self._fh.seek(self._offset)

        return self._fh

    def _update_offset_file(self):
        """
        Update the offset file with the current inode and offset.
        """
        offset = self._filehandle().tell()
        inode = stat(self.filename).st_ino
        fh = open(self._offset_file, "w")
        fh.write("%s\n%s\n" % (inode, offset))
        fh.close()

    def _determine_rotated_logfile(self):
        """
        We suspect the logfile has been rotated, so try to guess what the
        rotated filename is, and return it.
        """
        rotated_filename = self._check_rotated_filename_candidates()
        if (rotated_filename and exists(rotated_filename) and
            stat(rotated_filename).st_ino == self._offset_file_inode):
            return rotated_filename
        else:
            return None

    def _check_rotated_filename_candidates(self):
        """
        Check for various rotated logfile filename patterns and return the first
        match we find.
        """
        # savelog(8)
        candidate = "%s.0" % self.filename
        if (exists(candidate) and exists("%s.1.gz" % self.filename) and
            (stat(candidate).st_mtime > stat("%s.1.gz" % self.filename).st_mtime)):
            return candidate

        # logrotate(8)
        candidate = "%s.1" % self.filename
        if exists(candidate):
            return candidate

        # dateext rotation scheme
        candidates = glob.glob("%s-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]" % self.filename)
        if candidates:
            candidates.sort()
            return candidates[-1] # return most recent

        # no match
        return None

def main():
    # command-line parsing
    cmdline = OptionParser(usage="usage: %prog [options] logfile",
        description="Print log file lines that have not been read.")
    cmdline.add_option("--offset-file", "-o", action="store",
        help="File to which offset data is written (default: <logfile>.offset).")
    cmdline.add_option("--paranoid", "-p", action="store_true",
        help="Update the offset file every time we read a line (as opposed to"
             " only when we reach the end of the file).")

    options, args = cmdline.parse_args()

    if (len(args) != 1):
        cmdline.error("Please provide a logfile to read.")

    pygtail = Pygtail(args[0], 
                      offset_file=options.offset_file, 
                      paranoid=options.paranoid)
    for line in pygtail:
        sys.stdout.write(line)


if __name__ == "__main__":
    main()


pygtail/core.py
#!/usr/bin/python -tt
# -*- coding: utf-8 -*-

# pygtail - a python "port" of logtail2
# Copyright (C) 2011 Brad Greenlee <brad@footle.org>
#
# Derived from logcheck <http://logcheck.org>
# Copyright (C) 2003 Jonathan Middleton <jjm@ixtab.org.uk>
# Copyright (C) 2001 Paul Slootman <paul@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from __future__ import print_function
from os import fstat, stat
import os
from os.path import exists, getsize
import sys
import glob
import gzip
import io
from optparse import OptionParser

__version__ = '0.14.0'


PY3 = sys.version_info[0] == 3

if PY3:
    text_type = str
else:
    text_type = unicode


def force_text(s, encoding='utf-8', errors='strict'):
    if isinstance(s, text_type):
        return s
    return s.decode(encoding, errors)


class Offset:
    """Data-class to store file-offsets"""

    def __init__(self, counter, inode, offset):
        self.counter = counter
        self.inode = inode
        self.offset = offset

    def __eq__(self, other):
        return self.counter == other.counter and self.offset == other.offset

    def __lt__(self, other):
        return self.counter < other.counter or (
            self.counter == other.counter and self.offset < other.offset
        )

    def __le__(self, other):
        return self.__lt__(other) or self.__eq__(other)

    def __gt__(self, other):
        return not self.__le__(other)

    def __ge__(self, other):
        return not self.__le__(other) or self.__eq__(other)

    def __repr__(self):
        return "Offset(counter=%d, inode=%d, offset=%d" % (
            self.counter,
            self.inode,
            self.offset,
        )


class PygtailIteratorWithOffsets:
    def __init__(self, pygtail):
        self._pygtail = pygtail

    def __next__(self):
        return self.next()

    def __iter__(self):
        return self

    def next(self):
        next_line = self._pygtail.next()
        offset = self._pygtail._filehandle().tell()
        inode = fstat(self._pygtail._filehandle().fileno()).st_ino
        counter = self._pygtail._counter
        offset_instance = Offset(counter, inode, offset)
        return next_line, offset_instance


class Pygtail(object):
    """
    Creates an iterable object that returns only unread lines.

    Keyword arguments:
    offset_file   File to which offset data is written (default: <logfile>.offset).
    paranoid      Update the offset file every time we read a line (as opposed to
                  only when we reach the end of the file (default: False))
    every_n       Update the offset file every n'th line (as opposed to only when
                  we reach the end of the file (default: 0))
    on_update     Execute this function when offset data is written (default False)
    copytruncate  Support copytruncate-style log rotation (default: True)
    log_patterns  List of custom rotated log patterns to match (default: None)
    full_lines    Only log when line ends in a newline `\n` (default: False)
    save_on_end   Automatically save the offset once the end of the file is reached (default: True)
    """
    def __init__(self, filename, offset_file=None, paranoid=False, copytruncate=True,
                 every_n=0, on_update=False, read_from_end=False, log_patterns=None, full_lines=False,
                 save_on_end=True, encoding=None):
        self.filename = filename
        self.paranoid = paranoid
        self.every_n = every_n
        self.on_update = on_update
        self.copytruncate = copytruncate
        self.read_from_end = read_from_end
        self.log_patterns = log_patterns
        self.full_lines = full_lines
        self.save_on_end = save_on_end
        self.encoding = encoding
        self.offset_file = offset_file or "%s.offset" % self.filename
        self.offset_file_inode = 0
        self.offset = 0
        self.since_update = 0
        self.fh = None
        self.rotated_logfile = None
        self._counter = 0

        # if offset file exists and non-empty, open and parse it
        if exists(self.offset_file) and getsize(self.offset_file):
            offset_fh = open(self.offset_file, "r")
            (self.offset_file_inode, self.offset) = \
                [int(line.strip()) for line in offset_fh]
            offset_fh.close()
            if self.offset_file_inode != stat(self.filename).st_ino or \
                    stat(self.filename).st_size < self.offset:
                # The inode has changed or filesize has reduced so the file
                # might have been rotated.
                # Look for the rotated file and process that if we find it.
                self.rotated_logfile = self._determine_rotated_logfile()
                # If copytruncate is enabled and we can't find the rotated logfile, all we can do is reset.
                if self.copytruncate and self.rotated_logfile is None:
                    sys.stderr.write("[pygtail] [WARN] log file was rotated to unknown location. Resetting.\n")
                    self.offset = 0
                    self.update_offset_file()


    def __del__(self):
        if self._filehandle():
            self._filehandle().close()

    def __iter__(self):
        return self

    def next(self):
        """
        Return the next line in the file, updating the offset.
        """
        try:
            line = self._get_next_line()
        except StopIteration:
            # we've reached the end of the file; if we're processing the
            # rotated log file or the file has been renamed, we can continue with the actual file; otherwise
            # update the offset file
            if self._is_new_file():
                self.rotated_logfile = None
                self.fh.close()
                self.offset = 0
                # open up current logfile and continue
                try:
                    line = self._get_next_line()
                except StopIteration:  # oops, empty file
                    if self.save_on_end:
                        self.update_offset_file()
                    raise
            else:
                if self.save_on_end:
                    self.update_offset_file()
                raise

        if self.paranoid:
            self.update_offset_file()
        elif self.every_n and self.every_n <= self.since_update:
            self.update_offset_file()

        return line

    def with_offsets(self):
        """Returns an iterator that yields lines with their internal offset state"""
        return PygtailIteratorWithOffsets(self)

    def __next__(self):
        """`__next__` is the Python 3 version of `next`"""
        return self.next()

    def readlines(self):
        """
        Read in all unread lines and return them as a list.
        """
        return [line for line in self]

    def read(self):
        """
        Read in all unread lines and return them as a single string.
        """
        lines = self.readlines()
        if lines:
            try:
                return ''.join(lines)
            except TypeError:
                return ''.join(force_text(line) for line in lines)
        else:
            return None

    def _is_closed(self):
        if not self.fh:
            return True
        try:
            return self.fh.closed
        except AttributeError:
            if isinstance(self.fh, gzip.GzipFile):
                # python 2.6
                return self.fh.fileobj is None
            else:
                raise

    def _filehandle(self):
        """
        Return a filehandle to the file being tailed, with the position set
        to the current offset.
        """
        if not self.fh or self._is_closed():
            self._counter += 1
            filename = self.rotated_logfile or self.filename
            if filename.endswith('.gz'):
                self.fh = gzip.open(filename, 'r')
            elif PY3:
                self.fh = open(filename, "r", 1, encoding=self.encoding)
            else:
                self.fh = io.open(filename, "r", 1, encoding=self.encoding)
            if self.read_from_end and not exists(self.offset_file):
                self.fh.seek(0, os.SEEK_END)
            else:
                self.fh.seek(self.offset)

        return self.fh

    def update_offset_file(self):
        """
        Update the offset file with the current inode and offset.
        """
        if self.on_update:
            self.on_update()
        offset = self._filehandle().tell()
        inode = fstat(self._filehandle().fileno()).st_ino
        fh = open(self.offset_file, "w")
        fh.write("%s\n%s\n" % (inode, offset))
        fh.close()
        self.since_update = 0

    def write_offset_to_file(self, offset):
        """Writes an `Offset` to the offset file"""
        if self.on_update:
            self.on_update()
        fh = open(self.offset_file, "w")
        fh.write("%s\n%s\n" % (offset.inode, offset.offset))
        fh.close()

    def _determine_rotated_logfile(self):
        """
        We suspect the logfile has been rotated, so try to guess what the
        rotated filename is, and return it.
        """
        rotated_filename = self._check_rotated_filename_candidates()
        if rotated_filename and exists(rotated_filename):
            if stat(rotated_filename).st_ino == self.offset_file_inode:
                return rotated_filename

            # if the inode hasn't changed, then the file shrank; this is expected with copytruncate,
            # otherwise print a warning
            if stat(self.filename).st_ino == self.offset_file_inode:
                if self.copytruncate:
                    return rotated_filename
                else:
                    sys.stderr.write(
                        "[pygtail] [WARN] file size of %s shrank, and copytruncate support is "
                        "disabled (expected at least %d bytes, was %d bytes).\n" %
                        (self.filename, self.offset, stat(self.filename).st_size))

        return None

    def _check_rotated_filename_candidates(self):
        """
        Check for various rotated logfile filename patterns and return the first
        match we find.
        """
        # savelog(8)
        candidate = "%s.0" % self.filename
        if (exists(candidate) and exists("%s.1.gz" % self.filename) and
            (stat(candidate).st_mtime > stat("%s.1.gz" % self.filename).st_mtime)):
            return candidate

        # logrotate(8)
        # with delaycompress
        candidate = "%s.1" % self.filename
        if exists(candidate):
            return candidate

        # without delaycompress
        candidate = "%s.1.gz" % self.filename
        if exists(candidate):
            return candidate

        rotated_filename_patterns = [
            # logrotate dateext rotation scheme - `dateformat -%Y%m%d` + with `delaycompress`
            "%s-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]",
            # logrotate dateext rotation scheme - `dateformat -%Y%m%d` + without `delaycompress`
            "%s-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9].gz",
            # logrotate dateext rotation scheme - `dateformat -%Y%m%d-%s` + with `delaycompress`
            "%s-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]",
            # logrotate dateext rotation scheme - `dateformat -%Y%m%d-%s` + without `delaycompress`
            "%s-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9].gz",
            # for TimedRotatingFileHandler
            "%s.[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9]",
        ]
        if self.log_patterns:
            rotated_filename_patterns.extend(self.log_patterns)

        # break into directory and filename components to support cases where the
        # the file is prepended as part of rotation
        file_dir, rel_filename = os.path.split(self.filename)
        for rotated_filename_pattern in rotated_filename_patterns:
            candidates = glob.glob(os.path.join(file_dir, rotated_filename_pattern % rel_filename))
            if candidates:
                candidates.sort()
                return candidates[-1]  # return most recent

        # no match
        return None

    def _is_new_file(self):
        # Processing rotated logfile or at the end of current file which has been renamed
        return self.rotated_logfile or \
               self._filehandle().tell() == fstat(self._filehandle().fileno()).st_size and \
               fstat(self._filehandle().fileno()).st_ino != stat(self.filename).st_ino

    def _get_next_line(self):
        curr_offset = self._filehandle().tell()
        line = self._filehandle().readline()
        if self.full_lines:
            if not line.endswith('\n'):
                self._filehandle().seek(curr_offset)
                raise StopIteration
        if not line:
            raise StopIteration
        self.since_update += 1
        return line


def main():
    # command-line parsing
    cmdline = OptionParser(usage="usage: %prog [options] logfile",
        description="Print log file lines that have not been read.")
    cmdline.add_option("--offset-file", "-o", action="store",
        help="File to which offset data is written (default: <logfile>.offset).")
    cmdline.add_option("--paranoid", "-p", action="store_true",
        help="Update the offset file every time we read a line (as opposed to"
             " only when we reach the end of the file).")
    cmdline.add_option("--every-n", "-n", action="store",
        help="Update the offset file every n'th time we read a line (as opposed to"
             " only when we reach the end of the file).")
    cmdline.add_option("--no-copytruncate", action="store_true",
        help="Don't support copytruncate-style log rotation. Instead, if the log file"
             " shrinks, print a warning.")
    cmdline.add_option("--read-from-end", action="store_true",
        help="Read log file from the end if offset file is missing. Useful for large files.")
    cmdline.add_option("--log-pattern", action="append",
        help="Custom log rotation glob pattern. Use %s to represent the original filename."
             " You may use this multiple times to provide multiple patterns.")
    cmdline.add_option("--full_lines", action="store_true",
                       help="Only log when line ends in a newline (\\n)")
    cmdline.add_option("--encoding", action="store",
        help="Encoding to use for reading files (default: system encoding)")
    cmdline.add_option("--version", action="store_true",
        help="Print version and exit.")

    options, args = cmdline.parse_args()

    if options.version:
        print("pygtail version", __version__)
        sys.exit(0)

    if (len(args) != 1):
        cmdline.error("Please provide a logfile to read.")

    if options.every_n:
        options.every_n = int(options.every_n)
    pygtail = Pygtail(args[0],
                      offset_file=options.offset_file,
                      paranoid=options.paranoid,
                      every_n=options.every_n,
                      copytruncate=not options.no_copytruncate,
                      read_from_end=options.read_from_end,
                      log_patterns=options.log_pattern,
                      full_lines=options.full_lines,
                      encoding=options.encoding)

    for line in pygtail:
        sys.stdout.write(line)


if __name__ == "__main__":
    main()
